bridge/irc_listener.go

```diff
@@ -54,37 +58,74 @@ func (i *ircListener) nickTrackNick(event *irc.Event) {
 	}
 }
 
-// From irc_nicktrack.go.
-func (i *ircListener) nickTrackQuit(e *irc.Event) {
-	for k := range i.Connection.Channels {
-		delete(i.Channels[k].Users, e.Nick)
+func userOnChannelFix(user string, channel irc.Channel) bool {
+	if _, ok := channel.Users[user]; ok {
+		return true
 	}
-}
 
-func (i *ircListener) OnJoinQuitSettingChange() {
-	// Clear Nicktrack QUIT callback as it races with this
-	i.ClearCallback("QUIT")
-	i.ClearCallback("NICK")
-	i.AddCallback("NICK", i.nickTrackNick)
-
-	// If remove callbacks...
-	if !i.bridge.Config.ShowJoinQuit {
-		for event, id := range i.joinQuitCallbacks {
-			i.RemoveCallback(event, id) // note that QUIT was already removed above
+	// work around nicks being prefixed with mode characters for some reason
+	for _, c := range "!$~&@%+" {
+		if _, ok := channel.Users[string(c)+user]; ok {
+			return true
 		}
+	}
 
-		// Add back Nicktrack QUIT since it was removed
-		i.AddCallback("QUIT", i.nickTrackQuit)
+	return false
+}
+
+func (i *ircListener) OnNickRelayToDiscord(event *irc.Event) {
+	// Ignored hostmasks
+	if i.bridge.ircManager.isIgnoredHostmask(event.Source) {
 		return
 	}
 
-	callbacks := []string{"JOIN", "PART", "QUIT", "KICK"}
-	cbs := make(map[string]int, len(callbacks))
-	for _, cb := range callbacks {
-		i.AddCallback(cb, i.OnJoinQuitCallback)
+	oldNick := event.Nick
+	newNick := event.Message()
+
+	msg := IRCMessage{
+		Username: "",
+		Message:  fmt.Sprintf("_%s changed nick to %s_", oldNick, newNick),
 	}
 
-	i.joinQuitCallbacks = cbs
+	for _, m := range i.bridge.mappings {
+		channel := m.IRCChannel
+		channelObj, ok := i.Connection.Channels[channel]
+		if !ok {
+			continue
+		}
+
+		if !userOnChannelFix(newNick, channelObj) {
+			continue
+		}
+
+		msg.IRCChannel = channel
+		i.bridge.discordMessagesChan <- msg
+	}
+}
+
+// From irc_nicktrack.go.
+func (i *ircListener) nickTrackQuit(e *irc.Event) {
+	delete(i.bridge.ircManager.puppetNicks, e.Nick)
+}
+
+func (i *ircListener) OnJoinQuitSettingChange() {
+	// always remove our listener callbacks
+	for ev, id := range i.listenerCallbackIDs {
+		if id != 0 {
+			i.RemoveCallback(ev, id)
+			i.listenerCallbackIDs[ev] = 0
+		}
```

Comment:
```suggestion
		i.RemoveCallback(ev, id)
		delete(i.listenerCallbackIDs, ev)
```

(Also, `delete` no-ops if there is no such key.)